Numerically approximate arc perimeter

tomcur opened the pull request:
This is on top of PR https://github.com/linebender/kurbo/pull/378.

This implementation uses Carlson symmetric forms (https://arxiv.org/abs/math/9409227v1) of incomplete elliptic integrals of the second kind to numerically approximate elliptical arc lengths. These forms have nice computational properties allowing quick convergence of the approximations (each iteration reduces the error bound by 4⁶). Boost and Gnu Scientific Library are some other projects using them.

For the complete elliptic parts of the arc, the existing complete elliptic integral approximation (#383) is used.

waywardmonkeys committed:
Impl `ParamCurve`, `ParamCurveArclen` for `Arc`

This uses an approximation of the arc length using beziers as the
analytic solution is quite involved.

tomcur committed:
Numerically approximate ParamCurveArclen

waywardmonkeys commented:
I've no idea about the math here, but what Raph and I had discussed (somewhat) for #378 is that I would fix the one part that is broken and it could land and then a follow up (like this one?) would replace the current math with better math for the `arclen` operation.

tomcur commented:
This is based on top of your PR, changing `arclen` only. I'll rebase on top of your future fix , and rebase on main once your PR is merged.

The `arclen` implementation here needs more work, but that can proceed independently from fixes in #378.

tomcur committed:
Calculate for ellipse upper half, simplify arccos(a cos(x)/a)

Also add some more tests

tomcur committed:
Correctly wrap around

tomcur committed:
Fix negative sweep angle normalization

tomcur committed:
Add check comparing with bezier perimeter result

tomcur committed:
Improve docs

tomcur committed:
Improve readability

tomcur committed:
Clippy

tomcur committed:
Improve docs

tomcur commented:
Simplified a bit and hopefully all edge cases are now correct.

`arclen`'s `accuracy` param is not handled yet, but the Carlson approximations can be iterated to arbitrary precision. The easiest is to guarantee a relative `accuracy`, but an absolute `accuracy` might also be possible with some careful calculations.

tomcur committed:
Probable performance improvements

tomcur committed:
Fix wording

tomcur committed:
Division instead of `powi`

tomcur committed:
Start angle is modulo half turns

tomcur committed:
trunc is faster than floor

tomcur committed:
Reduce number of calls to integral calculation

tomcur commented:
I've optimized a bit, the implementation now requires three calls to the elliptic integrals, one of which is actually a complete integral and could be optimized with a quadratically converging approximation (which I suggest we leave for a later PR, that's also relevant for the `Ellipse` shape).

tomcur committed:
Provide rem_euclid in FloatFuncs

tomcur committed:
clippy: group digits

tomcur committed:
Fix erroneous agm_elliptic_perimeter if radius y > radius x

This doesn't (or doesn't easily?) occur for ellipses, as they instead
get a rotation when constructed with radius y > radius x.

tomcur committed:
Special-case complete elliptic perimeter

tomcur committed:
Lift calculation to compile-time

tomcur committed:
Don't calculate radii again

tomcur committed:
Improve wording

tomcur committed:
Add debug asserts

tomcur committed:
Merge sqrts

tomcur committed:
Calculate incomplete integrals to absolute accuracy

tomcur committed:
Add changelog

tomcur committed:
Remove comment

tomcur commented:
I need to go over the math once more when I'm fresh, but the relatively stringent tests pass, while the number of iterations per incomplete elliptic integral evaluation is around 3 or 4.

tomcur commented on src/arc.rs:
```diff
@@ -171,6 +177,118 @@ fn rotate_pt(pt: Vec2, angle: f64) -> Vec2 {
+        // that range.
+        let mut arclen = 0.;
+
+        // The available accuracy (tolerance) is distributed over the calculation of the two
+        // incomplete and one complete elliptic integrals.
```
The distribution of the accuracy tolerance is not very clever, but I'd rather not spend too many clock cycles on it, when those cycles could be spent iterating the convergence loops.

The worst case is when the number of complete quarter turns, evaluated as complete elliptic integrals, is very high, as the tolerance available to it decreases linearly in the number of quarter turns.

tomcur committed:
Group operations

tomcur committed:
Recip

tomcur commented on src/arc.rs:
```diff
@@ -223,6 +341,206 @@ impl Mul<Arc> for Affine {
+    let mut r = (accuracy / 3.).powf(2. / 3.) * e.powi(-4);
+
+    loop {
+        if 1. <= r * a.powi(5) * (1. - e / a) {
+            break;
+        }
```
This bound checking is a non-trivial amount of the operations per iteration, and while I think this is fine, there's an interesting alternative. Iterating to a relative error is cheaper (the formula is in "Numerical computation of real or complex elliptic integrals"), perhaps calculating a conservative upper bound before the loop starts and basing a relative error bound on that is also a possibility.

dominikh commented:
> This does still require #378 to be merged first.

Good news!
